design/14_tinkerbell_crd_refactor/proposal.md

```diff
@@ -0,0 +1,462 @@
+type StorageDevice string
+
+// KernelParam defines an atomic kernel parameter that will be passed to the OSIE.
+type KernelParam string
+```
+
+#### `OSIE`
+
+`OSIE` is a new CRD. It exists to ensure OSIE URLs can be re-used and easily updated across `Hardware` instances.
+
+```go
+// OSIE describes and Operating System Initialization Environment. It is used by Tinkerbell
+// to provision machines and should launch the Tink Worker component.
+type OSIE struct {
+	Spec OSIESpec
+}
+
+type OSIESpec struct {
+	// KernelURL is a URL to a kernel image.
+	KernelURL string
+
+	// InitrdURL is a URL to an initrd image.
+	InitrdURL string
+}
+```
+
+#### `Template`
+
+```go
+// Template defines a set of actions to be run on a target machine. The template is rendered
+// prior to execution where it is exposed to Hardware and user defined data. All fields within
+// TemplateSpec may contain template values. See https://pkg.go.dev/text/template for more details.
+type Template struct {
+	Spec TemplateSpec
+}
+
+type TemplateSpec struct {
+	// Actions defines the set of actions to be run on a target machine. Actions are run sequentially
+	// in the order they are specified. At least 1 action must be specified. Names of actions
+	// must be unique within a Template.
+	Actions []Action
+
+	// Volumes to be mounted on all actions. If an action specifies the same volume it will take
+	// precedence.
+	// Optional.
+	Volumes []Volume
+
+	// Env defines environment variables to be available in all actions. If an action specifies
+	// the same environment variable it will take precedence.
+	// Optional.
+	Env map[string]string
+}
+
+// Action defines an individual action to be run on a target machine.
+type Action struct {
+	// Name is a unique name for the action.
+	Name ActionName
+
+	// Image is an OCI image.
+	Image string
+
+	// Command defines the command to use when launching the image.
+	// Optional.
+	Command string
+
+	// Args are a set of arguments to be passed to the container on launch.
+	Args []string
+
+	// Env defines environment variables used when launching the container.
+	// Optional.
+	Env map[string]string
+
+		// Volumes defines the volumes to mount into the container.
+	// Optional.
+	Volumes []Volume
+
+	// NetworkNamespace defines the network namespace to run the container in. This enables access 
+	// to the host network namespace.
+	// See https://man7.org/linux/man-pages/man7/namespaces.7.html.
+	NetworkNamespace string
+}
+
+// Volume is a specification for mounting a volume in an action. Volumes take the form
+// {SRC-VOLUME-NAME | SRC-HOST-DIR}:TGT-CONTAINER-DIR:OPTIONS. When specifying a VOLUME-NAME that 
+// does not exist it will be created for you.
+//
+// Examples
+//
+// Read-only bind mount bound to /data
+//   /etc/data:/data:ro
+//
+// Writable volume name bound to /data
+//   shared_volume:/data
+//
+// See https://docs.docker.com/storage/volumes/ for additional details
+type Volume string
```

Comment:
For Volume I would assume that file system based volumes is not enough especially in k8s context. What about have a new volume of a type struct that we have in there different volume targets? e.g: volume from a file path, configmap, secret or source url.

Reply:
The volume doesn't relate to Kubernetes volumes; it pertains to Actions run on the OSIE.

Reply:
I see, so this should be just reflected just to the docker container and should only be running on the node(OISE) directly.